FirebaseStorage/Sources/Internal/StorageTokenAuthorizer.swift

```diff
@@ -0,0 +1,126 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+import FirebaseAppCheckInterop
+import FirebaseAuthInterop
+import FirebaseCore
+@_implementationOnly import FirebaseCoreExtension
+
+#if COCOAPODS
+  import GTMSessionFetcher
+#else
+  import GTMSessionFetcherCore
+#endif
+//
+//
+internal class StorageTokenAuthorizer: NSObject, GTMSessionFetcherAuthorizer {
+  func authorizeRequest(_ request: NSMutableURLRequest?,
+                        completionHandler handler: @escaping (Error?) -> Void) {
+    // Set version header on each request
+    let versionString = "ios/\(FirebaseVersion())"
+    request?.setValue(versionString, forHTTPHeaderField: "x-firebase-storage-version")
+
+    // Set GMP ID on each request
+    request?.setValue(googleAppID, forHTTPHeaderField: "x-firebase-gmpid")
+
+    var tokenError: NSError?
+    let callbackQueue = fetcherService.callbackQueue ?? DispatchQueue.main
+    let fetchTokenGroup = DispatchGroup()
+    if let auth = auth {
+      fetchTokenGroup.enter()
+      auth.getToken(forcingRefresh: false) { token, error in
+        if let error = error as? NSError {
+          var errorDictionary = error.userInfo
+          errorDictionary["ResponseErrorDomain"] = error.domain
+          errorDictionary["ResponseErrorCode"] = error.code
+          errorDictionary[NSLocalizedDescriptionKey] =
+            "User is not authenticated, please authenticate" +
+            " using Firebase Authentication and try again."
+          tokenError = NSError(domain: "FIRStorageErrorDomain",
+                               code: StorageErrorCode.unauthenticated.rawValue,
+                               userInfo: errorDictionary)
+        } else if let token = token {
+          let firebaseToken = "Firebase \(token)"
+          request?.setValue(firebaseToken, forHTTPHeaderField: "Authorization")
+        }
+        fetchTokenGroup.leave()
+      }
+    }
+    if let appCheck = appCheck {
+      fetchTokenGroup.enter()
+      appCheck.getToken(forcingRefresh: false) { tokenResult in
+        request?.setValue(tokenResult.token, forHTTPHeaderField: "X-Firebase-AppCheck")
+
+        if let error = tokenResult.error {
+          // TODO: Define better way to use FIRLogger from Swift.
+          FIRLogDebug1(
+            "[FirebaseStorage]",
+            "I-STR000001",
+            "Failed to fetch AppCheck token. Error: %@",
+            error
+          )
+        }
+        fetchTokenGroup.leave()
+      }
+    }
+    fetchTokenGroup.notify(queue: callbackQueue) {
+      handler(tokenError)
+    }
+  }
+
+  func authorizeRequest(_ request: NSMutableURLRequest?, delegate: Any, didFinish sel: Selector) {
```

Comment:
Can this method be removed?

Reply:
It's needed for the protocol. See the referenced GTMSessionFetcher PR for details.